api/src/main/java/org/pmiops/workbench/api/CohortsController.java

```diff
@@ -136,28 +127,61 @@ public void setUserProvider(Provider<User> userProvider) {
-    dbCohort.setVersion(1);
+
+    org.pmiops.workbench.db.model.Cohort newCohort = cohortFactory.createCohort(cohort, userProvider.get(), workspace.getWorkspaceId());
+
+    if (cohortDao.findCohortByNameAndWorkspaceId(newCohort.getName(), workspace.getWorkspaceId()) != null) {
+      throw new BadRequestException(String.format("Cohort \"/%s/%s/%s\" already exists.",
+              workspace.getWorkspaceNamespace(), workspace.getWorkspaceId(), newCohort.getName()));
+    }
+
     try {
       // TODO Make this a pre-check within a transaction?
-      dbCohort = cohortDao.save(dbCohort);
-      userRecentResourceService.updateCohortEntry(workspace.getWorkspaceId(), userProvider.get().getUserId(), dbCohort.getCohortId(), now);
+      newCohort = cohortDao.save(newCohort);
+      userRecentResourceService.updateCohortEntry(workspace.getWorkspaceId(),
+              userProvider.get().getUserId(),
+              newCohort.getCohortId(),
+              new Timestamp(clock.instant().toEpochMilli()));
     } catch (DataIntegrityViolationException e) {
       // TODO The exception message doesn't show up anywhere; neither logged nor returned to the
       // client by Spring (the client gets a default reason string).
-      throw new BadRequestException(String.format(
-          "Cohort \"/%s/%s/%d\" already exists.",
-          workspaceNamespace, workspaceId, dbCohort.getCohortId()));
+      throw new ServerErrorException(
+              String.format("Could not save Cohort (\"/%s/%s/%s\")", workspace.getWorkspaceNamespace(), workspace.getWorkspaceId(), newCohort.getName()),
+              e
+      );
     }
-    return ResponseEntity.ok(TO_CLIENT_COHORT.apply(dbCohort));
+
+    return ResponseEntity.ok(TO_CLIENT_COHORT.apply(newCohort));
+  }
+
+  @Override
+  public ResponseEntity<Cohort> duplicateCohort(String workspaceNamespace, String workspaceId, Long cohortId) {
+    workspaceService.enforceWorkspaceAccessLevel(workspaceNamespace, workspaceId, WorkspaceAccessLevel.WRITER);
+    Workspace workspace = workspaceService.getRequired(workspaceNamespace, workspaceId);
+
+    org.pmiops.workbench.db.model.Cohort fromCohort = getDbCohort(workspaceNamespace, workspaceId, cohortId);
+    org.pmiops.workbench.db.model.Cohort newCohort = cohortFactory.duplicateCohort(fromCohort, userProvider.get());
+
+    if (cohortDao.findCohortByNameAndWorkspaceId(newCohort.getName(), workspace.getWorkspaceId()) != null) {
+      throw new BadRequestException(String.format("Cohort \"/%s/%s/%s\" already exists.",
+              workspace.getWorkspaceNamespace(), workspace.getWorkspaceId(), newCohort.getName()));
+    }
+
+    try {
+      newCohort = cohortDao.save(newCohort);
+      userRecentResourceService.updateCohortEntry(workspace.getWorkspaceId(),
+              userProvider.get().getUserId(),
+              newCohort.getCohortId(),
+              new Timestamp(clock.instant().toEpochMilli()));
+    } catch (Exception e) {
+      throw new ServerErrorException(
```

Comment:
Opt-nit: Do we know what sorts of errors we could be seeing here? It would be good to throw something more descriptive than a 500 if we can.

Reply:
Not really sure unfortunately.

That said, I believe it is an improvement over what was happening previously where all `DataIntegrityViolationException` would get caught as the naming conflict exception message and any other errors get bubbled up to a 500 which is what is happening here, just more explicitly.

I could get rid of the try/catch and it should return 500s in the same conditions but we would lose the exception message.